becquerel/core/spectrum.py

```diff
@@ -722,9 +894,81 @@ def downsample(self, f, handle_livetime=None):
         old_counts = self.counts_vals.astype(int)
         new_counts = np.random.binomial(old_counts, 1. / f)
 
-        return Spectrum(counts=new_counts,
-                        bin_edges_kev=self.bin_edges_kev,
-                        livetime=new_livetime)
+        if self.is_calibrated:
+            return Spectrum(counts=new_counts,
+                            bin_edges_kev=self.bin_edges_kev,
+                            livetime=new_livetime)
+        else:
+            return Spectrum(counts=new_counts,
+                            bin_edges_raw=self.bin_edges_raw,
+                            livetime=new_livetime)
+
+    def has_uniform_bins(self, use_kev=True):
+        """Test whether the Spectrum has uniform binning.
+
+        This is possibly the best way to test for uniform binning within float
+        precision, after https://github.com/0cjs/py-allthesame. It may also be
+        useful to make this a member, not a method, that gets set at Spectrum
+        initialization.
+
+        Args:
+          use_kev: check bin_edges_kev if True, otherwise bin_edges_raw. This
+            is a rather unfortunate workaround necessary when keeping both the
+            raw and kev versions of the bin_edges
+
+        Raises:
+          SpectrumError: if use_kev=True but Spectrum is not calibrated
+
+        Returns:
+          True/False if all binwidths are equal to within np.finfo(float).eps
+        """
+
+        if use_kev and not self.is_calibrated:
+            raise SpectrumError('Cannot access energy bins with an ' +
+                                'uncalibrated Spectrum.')
+
+        bin_widths = self.bin_widths_kev if use_kev else self.bin_widths_raw
+        iterator = iter(bin_widths)
+        x0 = next(iterator, None)
+        for x in iterator:
+            if abs(x/x0 - 1.0) > EPS:
+                return False
+        return True
+
+    def find_bin(self, x, use_kev=True):
+        """Find the Spectrum bin that contains a value x.
+
+        If the Spectrum has uniform binning, then we just solve the linear
+        equation between bins and x-axis values. Otherwise, we use searchsorted
+        to bisect for the insertion point where x would fit in a list of bin
+        edges, then subtract 1 to get the index of the low edge.
+
+        Args:
+          x: value whose bin to find
+          use_kev: check bin_edges_kev if True, otherwise bin_edges_raw
+
+        Raises:
+          AssertionError: if x is outside the bin edges or equal to up edge
+          SpectrumError: if use_kev=True but Spectrum is not calibrated
+
+        Returns:
+          The integer bin number that contains x
```

Comment:
There's a subtle semantic issue with this method, and I would suggest renaming to `find_bin_index` or something of the sort. The term "bin" is a little ambiguous as to whether it is referring to the index of the bin or, e.g., the value of the bin center. Do others have the same confusion?

Reply:
I'm used to the ROOT convention where a "bin" is always an integer index, and methods exist to compute bin low/up edges or centers. See, e.g., [here](https://root.cern.ch/doc/master/classTH1.html#ae0895b66e993b9e9ec728d828b9336fe).

Reply:
This has been resolved.

Reply:
Now called `find_bin_index()`